qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,197 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Optional, Set, Union
+
+import sympy
+from attrs import frozen
+
+from qualtran import Bloq, Register, Signature
+from qualtran.bloqs.basic_gates import Toffoli
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class ModAdd(Bloq):
+    r"""An n-bit modular addition gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle \rightarrow |x\rangle|y + x mod p\rangle|p\rangle$ using $4n Toffoli$ gates.
```

Comment:
nit: `\mod` 

`$4n$ Toffoli`

please wrap to 100 characters

here and everywhere